sql/core/src/main/scala/org/apache/spark/sql/Column.scala

```diff
@@ -828,11 +828,11 @@ class Column(val expr: Expression) extends Logging {
    * @since 2.4.0
    */
   def isInCollection(values: scala.collection.Iterable[_]): Column = withExpr {
-    val hSet = values.toSet[Any]
-    if (hSet.size > SQLConf.get.optimizerInSetConversionThreshold) {
-      InSet(expr, hSet)
+    val exprValues = values.toSeq.map(lit(_).expr)
+    if (exprValues.size > SQLConf.get.optimizerInSetConversionThreshold) {
+      InSet(expr, exprValues.map(_.eval()).toSet)
     } else {
-      In(expr, values.toSeq.map(lit(_).expr))
+      In(expr, exprValues)
```

Comment:
So, this is caused by SPARK-29048 (Improve performance on Column.isInCollection() with a large size collection, https://github.com/apache/spark/pull/25754 ) and only affects 3.0.0, right?

Reply:
Correct

Reply:
Thanks for confirming, @MaxGekk .
cc @WeichenXu123 and @gatorsmile